neonvm/runner/containermgr.go

```diff
@@ -0,0 +1,364 @@
+			)
+			time.Sleep(retryInitEvery)
+		}
+	}
+
+	logger.Info(
+		fmt.Sprintf("Got CRI ID for %s container", runnerContainerName),
+		zap.String("containerID", criRunnerContainerID),
+	)
+
+	cinfo := containerInfo{
+		podID:        criPodID,
+		containerID:  criRunnerContainerID,
+		lastMilliCPU: atomic.Uint32{},
+	}
+
+	cinfo.listenForCPUChanges(context.TODO(), logger, crictl, int32(httpPort))
+}
+
+type crictlConfig struct {
+	endpoint string
+}
+
+func (c *crictlConfig) run(logger *zap.Logger, output any, args ...string) error {
+	actualArgs := []string{
+		"--runtime-endpoint",
+		c.endpoint,
+	}
+	actualArgs = append(actualArgs, args...)
+
+	logger.Info("running crictl", zap.Strings("args", actualArgs))
+
+	cmd := exec.Command("/usr/bin/crictl", actualArgs...)
+
+	stderr, err := os.OpenFile("/dev/stderr", os.O_RDWR, 0 /* unused */)
+	if err != nil {
+		panic(fmt.Errorf("failed to open /dev/stderr: %w", err))
+	}
+	cmd.Stderr = stderr
+
+	if output == nil {
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("failed to run command: %w", err)
+		}
+		return nil
+	} else {
+		out, err := cmd.Output()
+		if err != nil {
+			return fmt.Errorf("failed to run command: %w", err)
+		}
+
+		if err := json.Unmarshal(out, output); err != nil {
+			return fmt.Errorf("error parsing JSON: %w", err)
+		}
+		return nil
+	}
+}
+
+type containerInfo struct {
+	podID        string
+	containerID  string
+	lastMilliCPU atomic.Uint32
+}
+
+func (c *containerInfo) listenForCPUChanges(ctx context.Context, logger *zap.Logger, crictl *crictlConfig, port int32) {
+	mux := http.NewServeMux()
+	loggerHandlers := logger.Named("http-handlers")
+	cpuChangeLogger := loggerHandlers.Named("cpu_change")
+	mux.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		_ = r.Body.Close()
+		w.WriteHeader(200)
+	})
+	mux.HandleFunc("/cpu_change", func(w http.ResponseWriter, r *http.Request) {
+		c.handleCPUChange(cpuChangeLogger, crictl, w, r)
+	})
+	cpuCurrentLogger := loggerHandlers.Named("cpu_current")
+	mux.HandleFunc("/cpu_current", func(w http.ResponseWriter, r *http.Request) {
+		c.handleCPUCurrent(cpuCurrentLogger, crictl, w, r)
+	})
+	server := http.Server{
+		Addr:              fmt.Sprintf("0.0.0.0:%d", port),
+		Handler:           mux,
+		ReadTimeout:       5 * time.Second,
+		ReadHeaderTimeout: 5 * time.Second,
+		WriteTimeout:      5 * time.Second,
+	}
+	errChan := make(chan error)
+	go func() {
+		errChan <- server.ListenAndServe()
+	}()
+	select {
+	case err := <-errChan:
+		if errors.Is(err, http.ErrServerClosed) {
+			logger.Info("cpu_change server closed")
+		} else if err != nil {
+			logger.Fatal("cpu_change server exited with error", zap.Error(err))
+		}
+	case <-ctx.Done():
+		err := server.Shutdown(context.Background())
+		logger.Info("shut down cpu_change server", zap.Error(err))
+	}
+}
+
+func (c *containerInfo) handleCPUChange(logger *zap.Logger, crictl *crictlConfig, w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
```

Comment:
We can use `http.MethodPost` and it's nicer

Reply:
Copied from neonvm-runner, happy to address in a follow-up.

https://github.com/neondatabase/autoscaling/blob/32e407a6ba1ae49c341e9de898d62489914fff74/neonvm/runner/main.go#L718